# Hashing Algorithm

Hashing Vectorizer converts text into fixed-length numerical feature vectors using a hash function, eliminating the need to build and store a vocabulary. It is memory-efficient and well-suited for large-scale text processing.

In [1]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import HashingVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn import metrics
from sklearn.metrics import classification_report

In [2]:
dataset = pd.read_csv("news.csv").drop(columns=["Unnamed: 0"])

In [3]:
dataset

,title,text,label
0,You Can Smell Hillary’s Fear,"Daniel Greenfield, a Shillman Journalism Fello...",FAKE
1,Watch The Exact Moment Paul Ryan Committed Pol...,Google Pinterest Digg Linkedin Reddit Stumbleu...,FAKE
2,Kerry to go to Paris in gesture of sympathy,U.S. Secretary of State John F. Kerry said Mon...,REAL
3,Bernie supporters on Twitter erupt in anger ag...,"— Kaydee King (@KaydeeKing) November 9, 2016 T...",FAKE
4,The Battle of New York: Why This Primary Matters,It's primary day in New York and front-runners...,REAL
...,...,...,...
6330,State Department says it can't find emails fro...,The State Department told the Republican Natio...,REAL
6331,The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...,The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...,FAKE
6332,Anti-Trump Protesters Are Tools of the Oligarc...,Anti-Trump Protesters Are Tools of the Oligar...,FAKE
6333,"In Ethiopia, Obama seeks progress on peace, se...","ADDIS ABABA, Ethiopia —President Obama convene...",REAL


In [4]:
y=dataset["label"]

In [5]:
X_train, X_test, y_train, y_test = train_test_split(dataset['text'], y, test_size=0.33, random_state=53)

Here we no need to call fit() as it is not learning any vocabulary from the training data here. It simply hashes words directly.

In [7]:
hash_vectorizer = HashingVectorizer(stop_words='english',n_features=2**18,alternate_sign=False)
hash_train = hash_vectorizer.transform(X_train)
print(hash_train)
hash_test = hash_vectorizer.transform(X_test)

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 1118691 stored elements and shape (4244, 262144)>
  Coords	Values
  (1, 9728)	0.2672612419124244
  (1, 12144)	0.1336306209562122
  (1, 26381)	0.1336306209562122
  (1, 46353)	0.1336306209562122
  (1, 76282)	0.1336306209562122
  (1, 88275)	0.1336306209562122
  (1, 91609)	0.1336306209562122
  (1, 92622)	0.1336306209562122
  (1, 93434)	0.1336306209562122
  (1, 102563)	0.1336306209562122
  (1, 115574)	0.1336306209562122
  (1, 117992)	0.2672612419124244
  (1, 121388)	0.2672612419124244
  (1, 124604)	0.1336306209562122
  (1, 137783)	0.1336306209562122
  (1, 151171)	0.2672612419124244
  (1, 159607)	0.2672612419124244
  (1, 167229)	0.1336306209562122
  (1, 190636)	0.1336306209562122
  (1, 216568)	0.1336306209562122
  (1, 244285)	0.5345224838248488
  (1, 253040)	0.1336306209562122
  (1, 253074)	0.2672612419124244
  (2, 14361)	0.3333333333333333
  (2, 15347)	0.3333333333333333
  :	:
  (4243, 224972)	0.028261670947211076
  (4243, 225649

Every word is converted into a large integer using a mathematical formula.
word -- hash function -- modulo operation with no of columns(no. of features) -- if remainder is 0 it belongs to col 0.
feature_index = hash(word) % n_features

n_features = 2**18
each document has 262,144 columns.

A larger value reduces the chance that different words map to the same feature.

In [8]:
clf = MultinomialNB()

clf.fit(hash_train, y_train)
pred = clf.predict(hash_test)
score = metrics.accuracy_score(y_test, pred)
print("accuracy:   %0.3f" % score)

cm = metrics.confusion_matrix(y_test, pred, labels=['FAKE', 'REAL'])

accuracy:   0.857


In [9]:
from sklearn.metrics import classification_report

report=classification_report(y_test, pred)


In [10]:
print(report)

              precision    recall  f1-score   support

        FAKE       0.95      0.74      0.83      1008
        REAL       0.80      0.97      0.88      1083

    accuracy                           0.86      2091
   macro avg       0.88      0.85      0.85      2091
weighted avg       0.87      0.86      0.85      2091



In [11]:
dataset["text"][2]

'U.S. Secretary of State John F. Kerry said Monday that he will stop in Paris later this week, amid criticism that no top American officials attended Sunday’s unity march against terrorism.\n\nKerry said he expects to arrive in Paris Thursday evening, as he heads home after a week abroad. He said he will fly to France at the conclusion of a series of meetings scheduled for Thursday in Sofia, Bulgaria. He plans to meet the next day with Foreign Minister Laurent Fabius and President Francois Hollande, then return to Washington.\n\nThe visit by Kerry, who has family and childhood ties to the country and speaks fluent French, could address some of the criticism that the United States snubbed France in its darkest hour in many years.\n\nThe French press on Monday was filled with questions about why neither President Obama nor Kerry attended Sunday’s march, as about 40 leaders of other nations did. Obama was said to have stayed away because his own security needs can be taxing on a country, 

In [12]:
hash_vectorizer = HashingVectorizer(stop_words='english',n_features=2**18,alternate_sign=False)
hash_train = hash_vectorizer.transform(X_train)


In [13]:
X_train[[2]]

2    U.S. Secretary of State John F. Kerry said Mon...
Name: text, dtype: object

In [14]:
clf.predict(hash_train[[2]])

array(['REAL'], dtype='<U4')